datastore/datastore.go

```diff
@@ -0,0 +1,260 @@
+					}
+					failedLock.Unlock()
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return func(q query.Query, res query.Results, err error) (query.Results, error) {
+		if err != nil {
+			return res, err
+		}
+
+		k := datastore.NewKey(q.Prefix)
+
+		// not interested if this is not a query for providers of a particular peer
+		// we're looking for /providers/cid, not /providers (currently used in GC)
+		if !providersRoot.IsAncestorOf(k) || len(k.Namespaces()) < 2 {
+			return res, err
+		}
+
+		failedLock.Lock()
+		if failed.Has(k.String()) {
+			failedLock.Unlock()
+			recordFindProvsComplete(ctx, "discarded")
+			return res, err
+		}
+		failedLock.Unlock()
+
+		var count int
+		onAfterNextSync := func(r query.Result, ok bool) (query.Result, bool) {
+			if ok {
+				count++
+				return r, ok
+			}
+			if count > 0 { // query has ended and there were locally found records
+				recordFindProvsComplete(ctx, "local")
+				return r, ok
+			}
+
+			pendingLock.Lock()
+			isPending, _ := pending[k.String()]
+			if !isPending {
+				// send to the find provs queue, if channel is full then discard...
+				select {
+				case findC <- k:
+					pending[k.String()] = true
+					stats.Record(ctx, metrics.FindProvsQueueSize.M(1))
+				case <-ctx.Done():
+				default:
+					recordFindProvsComplete(ctx, "discarded")
+				}
+			}
+			pendingLock.Unlock()
+
+			return r, ok
+		}
+
+		return hres.NewResults(res, hropts.OnAfterNextSync(onAfterNextSync)), nil
+	}
+}
+
+func findProviders(ctx context.Context, findC chan datastore.Key, foundC chan findResult, getRouting GetRoutingFunc, opts Options) {
+	done := func(r findResult) {
+		select {
+		case foundC <- r:
+		case <-ctx.Done():
+		}
+	}
+
+	for {
+		select {
+		case k := <-findC:
+			cid, err := providerKeyToCID(k)
+			if err != nil {
+				done(findResult{key: k, err: fmt.Errorf("failed to create CID from provider record key: %w", err)})
+				continue
+			}
+
+			routing, addProvider, err := getRouting(cid)
+			if err != nil {
+				done(findResult{key: k, err: fmt.Errorf("failed to get routing for CID: %w", err)})
+				continue
+			}
+
+			status := "failed"
+			start := time.Now()
+			fctx, cancel := context.WithTimeout(ctx, opts.FindProvidersTimeout)
+
+			for ai := range routing.FindProvidersAsync(fctx, cid, opts.FindProvidersCount) {
+				addProvider(ctx, cid, ai.ID)
+				status = "succeeded"
+			}
+
+			cancel()
+			done(findResult{key: k, status: status, duration: time.Since(start) / 1e+9})
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+var errInvalidKeyNamespaces = fmt.Errorf("not enough namespaces in provider record key")
+
+func providerKeyToCID(k datastore.Key) (cid.Cid, error) {
+	nss := k.Namespaces()
+	if len(nss) < 2 {
+		return cid.Undef, errInvalidKeyNamespaces
+	}
+
+	b, err := base32.RawStdEncoding.DecodeString(nss[1])
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	_, c, err := cid.CidFromBytes(b)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	return c, nil
+}
```

Comment:
Don't we use this in more places (e.g. HTTP API)? Perhaps can be exported and reused, no?

Reply:
I did not realise - yes I’ll move it to a util or something that both can use!

Reply:
Hmm, just looked again I don't think we do this anywhere else. You did something similar here https://github.com/libp2p/go-libp2p-kad-dht/pull/493/files#diff-8c86329cd083a65dc37f07ce95b47f59R277-R283